cpp/src/io/text/bgzip_data_chunk_source.cu

```diff
@@ -0,0 +1,444 @@
+                                    _curr_block.d_decompressed_blocks.data() + _curr_block.read_pos,
+                                    read_size,
+                                    cudaMemcpyDeviceToDevice,
+                                    cuda_stream.value()));
+      // record the host-to-device copy, decompression and device copy
+      CUDF_CUDA_TRY(cudaEventRecord(_curr_block.event, cuda_stream.value()));
+      _curr_block.consume_bytes(read_size);
+      return std::make_unique<device_uvector_data_chunk>(std::move(data));
+    }
+    read_next_compressed_chunk(read_size /* - _curr_block.remaining_size()*/);
+    _prev_block.decompress(cuda_stream);
+    _curr_block.decompress(cuda_stream);
+    read_size = std::min(read_size, _prev_block.remaining_size() + _curr_block.remaining_size());
+    rmm::device_uvector<char> data(read_size, cuda_stream);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(data.data(),
+                                  _prev_block.d_decompressed_blocks.data() + _prev_block.read_pos,
+                                  _prev_block.remaining_size(),
+                                  cudaMemcpyDeviceToDevice,
+                                  cuda_stream.value()));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(data.data() + _prev_block.remaining_size(),
+                                  _curr_block.d_decompressed_blocks.data() + _curr_block.read_pos,
+                                  read_size - _prev_block.remaining_size(),
+                                  cudaMemcpyDeviceToDevice,
+                                  cuda_stream.value()));
+    // record the host-to-device copy, decompression and device copy
+    CUDF_CUDA_TRY(cudaEventRecord(_curr_block.event, cuda_stream.value()));
+    CUDF_CUDA_TRY(cudaEventRecord(_prev_block.event, cuda_stream.value()));
+    read_size -= _prev_block.remaining_size();
+    _prev_block.consume_bytes(_prev_block.remaining_size());
+    _curr_block.consume_bytes(read_size);
+    return std::make_unique<device_uvector_data_chunk>(std::move(data));
+  }
+
+ private:
+  std::unique_ptr<std::istream> _stream;
+  compressed_blocks _prev_block;
+  compressed_blocks _curr_block;
+  std::size_t _local_end;
+  std::size_t _compressed_pos;
+  std::size_t _compressed_end;
+};
+
+class bgzip_data_chunk_source : public data_chunk_source {
+ public:
+  bgzip_data_chunk_source(std::string filename, uint64_t virtual_begin, uint64_t virtual_end)
+    : _filename{std::move(filename)}, _virtual_begin{virtual_begin}, _virtual_end{virtual_end}
+  {
+  }
+  [[nodiscard]] std::unique_ptr<data_chunk_reader> create_reader() const override
+  {
+    return std::make_unique<bgzip_data_chunk_reader>(
+      std::make_unique<std::ifstream>(_filename, std::ifstream::in), _virtual_begin, _virtual_end);
+  }
+
+ private:
+  std::string _filename;
+  uint64_t _virtual_begin;
+  uint64_t _virtual_end;
+};
+
+}  // namespace
+
+std::unique_ptr<data_chunk_source> make_source_from_bgzip_file(std::string const& filename,
```

Comment:
Can we replace `std::string const&` with `std::string_view`?

Reply:
Not sure about this - std::ifstream doesn't seem to have a string_view constructor, and assuming the input is null-terminated seems dangerous.

Reply:
> std::ifstream doesn't seem to have a string_view constructor

Not sure I'm following the issue here, `string_view::data()` returns a char pointer that can be used to construct `ifstream`. 

It seems `filename` is used to initialize the `std::string _filename` member in `class bgzip_data_chunk_source`

Reply:
True, I missed the indirection via the data chunk source. `string_view::data()` can't be safely passed to the `ifstream` because its `const char*` constructor assumes a C style null-terminated string, while string_view may end with other characters.